runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/stableinput/BufferingDoFnRunner.java

```diff
@@ -143,15 +174,15 @@ public void checkpoint(long checkpointId) throws Exception {
     // We are about to get checkpointed. The elements buffered thus far
     // have to be added to the global CheckpointElement state which will
     // be used to emit elements later when this checkpoint is acknowledged.
-    addToBeAcknowledgedCheckpoint(checkpointId, currentStateId);
-    currentStateId = generateNewId();
-    currentBufferingElementsHandler = bufferingElementsHandlerFactory.get(currentStateId);
+    addToBeAcknowledgedCheckpoint(checkpointId, getStateIndex());
+    int newStateIndex = rotateAndGetStateIndex();
+    currentBufferingElementsHandler = bufferingElementsHandlerFactory.get(newStateIndex);
   }
 
   /** Should be called when a checkpoint is completed. */
   public void checkpointCompleted(long checkpointId) throws Exception {
```

Comment:
just to check if I understand this correctly:
 
- We are guaranteed that checkpoints complete in sequential order.
- We might get gaps in case of failed / expired checkpoints, so any completed checkpoint must complete all prior checkpoints.

Reply:
It's good that you bring this up because I wanted to add a comment to explain this. Since every checkpoint runs independently we are not guaranteed to complete in sequential order. In case a later checkpoint completes before an earlier one, we can safely process the elements of the earlier one. This method is guaranteed not to overlap because Flink uses a single thread to execute async task calls lie (a) triggering a checkpoint (b) completing a checkpoint. When an earlier checkpoint completes after a later one has already completed, we will do nothing here.

Reply:
I wonder whether we should make the flushing backwards-compatible for users who want to migrate.

Reply:
I don't have strong opinions about this. In this case I'd incline to bw incompatible change.